apps/src/studio/studio.js

```diff
@@ -1894,43 +1884,56 @@ Studio.reset = function(first) {
   for (i = 0; i < Studio.spriteCount; i++) {
     Studio.displaySprite(i);
   }
+  this.resetGoalSprites();
+  sortDrawOrder();
 
-  var goalAsset = skin.goal;
-  if (level.goalOverride) {
-    if (level.goalOverride.goalAnimation) {
-      goalAsset = skin[level.goalOverride.goalAnimation];
-    } else if (level.goalOverride.goalImage) {
-      goalAsset = skin[level.goalOverride.goalImage];
-    }
-  }
+  // A little flag for script-based code to consume.
+  Studio.levelRestarted = true;
 
+  // Reset whether level has succeeded.
+  Studio.succeededTime = null;
+};
+
+/**
+ * Move all goal sprites to their original positions, and reset their completion
+ * state, both visual and logical.
+ */
+Studio.resetGoalSprites = function () {
   Studio.touchAllGoalsEventFired = false;
-  for (i = 0; i < Studio.spriteGoals_.length; i++) {
+  for (var i = 0; i < Studio.spriteGoals_.length; i++) {
```

Comment:
Code smell: Any time you see yourself reusing a declaration of `i` for many loops, your function is too long!